### PR TITLE
Add CL simple concurrency test for template creation

### PR DESCRIPTION
**Pull request by sjug**
This tests exists to speed up tests like master vertical which have a large number of projects and large number of templates within them.

### CODE REVIEW

**Comment by sjug** [comment]
The new command to run this test looks like:
```
VIPERCONFIG=/home/user/content/pyconfigMasterVertScalePause.yaml ./_output/local/bin/linux/amd64/openshift-tests run-test "[Feature:Performance][Serial][Fast] Load cluster with templates [Suite:openshift/conformance/serial]"
```

There is a new config field that has threads defined:
```
provider: local
ClusterLoader:
  cleanup: true
  threads: 3
  projects:
```
Default thread value is 10 if unset.

One oddity that I don't understand is when the test completes:
```
...
route.route.openshift.io/route2 created
Worker DONE!
route.route.openshift.io/route2 created
Worker DONE!
route.route.openshift.io/route2 created
Worker DONE!
[AfterEach] [Top Level]
  /home/jugs/go/src/github.com/openshift/origin/_output/local/go/src/github.com/openshift/origin/test/extended/util/cli.go:108
STEP: Collecting events from namespace "c96336".
[AfterEach] [Top Level]
  /home/jugs/go/src/github.com/openshift/origin/_output/local/go/src/github.com/openshift/origin/test/extended/util/cli.go:108
STEP: Collecting events from namespace "c29724".
[AfterEach] [Top Level]
  /home/jugs/go/src/github.com/openshift/origin/_output/local/go/src/github.com/openshift/origin/test/extended/util/cli.go:108
STEP: Collecting events from namespace "c69001".
[AfterEach] [Top Level]
  /home/jugs/go/src/github.com/openshift/origin/_output/local/go/src/github.com/openshift/origin/vendor/k8s.io/kubernetes/test/e2e/framework/framework.go:150
Jun 12 10:48:41.357: INFO: Waiting up to 3m0s for all (but 100) nodes to be ready
[AfterEach] [Top Level]
  /home/jugs/go/src/github.com/openshift/origin/_output/local/go/src/github.com/openshift/origin/vendor/k8s.io/kubernetes/test/e2e/framework/framework.go:150
Jun 12 10:48:41.357: INFO: Waiting up to 3m0s for all (but 100) nodes to be ready
[AfterEach] [Top Level]
  /home/jugs/go/src/github.com/openshift/origin/_output/local/go/src/github.com/openshift/origin/vendor/k8s.io/kubernetes/test/e2e/framework/framework.go:150
Jun 12 10:48:41.358: INFO: Waiting up to 3m0s for all (but 100) nodes to be ready
Jun 12 10:48:41.360: INFO: Running AfterSuite actions on all nodes
Jun 12 10:48:41.360: INFO: Running AfterSuite actions on node 1
fail [github.com/openshift/origin/test/extended/util/cli.go:108]: You may only call AfterEach from within a Describe or Context
[user@host origin]$ echo $?
1
``

The test seems to fail?

File: test/extended/cluster/clt.go [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+			allArgs = append(allArgs, "--node-selector")
+			allArgs = append(allArgs, p.NodeSelector)
+		}
+		nsName := fmt.Sprintf("%s%d", p.Basename, p.Counter)
+		allArgs = append(allArgs, nsName)
+
+		// Create namespaces as defined in Cluster Loader config
+		err := oc.Run("adm", "new-project").Args(allArgs...).Execute()
+		if err != nil {
+			fmt.Printf("Error creating project: %v\n", err)
```
Comment (jmencak) [inline test/extended/cluster/clt.go]: Is there a reason to mix `fmt.Printf()` and `e2e.Logf()`?

File: test/extended/cluster/clt.go [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+		if project == nil {
+			e2e.Failf("Invalid config file.\nFile: %v", project)
+		}
+
+		numWorkers := ConfigContext.ClusterLoader.Threads
+		if numWorkers == 0 {
+			numWorkers = DefaultThreadCount
+		}
+
+		work := make(chan ProjectMeta, 1024)
```
Comment (jmencak) [inline test/extended/cluster/clt.go]: Why 1024?  Perhaps make is a descriptive constant?

File: test/extended/cluster/clt.go [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+				unit := ProjectMeta{j, p}
+				work <- unit
+			}
+		}
+
+		wg.Wait()
+	})
+})
+
+func clWorker(in <-chan ProjectMeta, wg *sync.WaitGroup) {
```
Comment (jmencak) [inline test/extended/cluster/clt.go]: This never returns? Goroutine leak?

File: test/extended/cluster/clt.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (sjug) [reply test/extended/cluster/clt.go]: Good spot, moved to consistent logging functions.

File: test/extended/cluster/clt.go (hunk 2 above)
Comment (sjug) [reply test/extended/cluster/clt.go]: I just thought that was a large number, what do you suggest?

File: test/extended/cluster/clt.go (hunk 3 above)
Comment (sjug) [reply test/extended/cluster/clt.go]: Should be fixed now, good find.

File: test/extended/cluster/clt.go (hunk 2 above)
Comment (jmencak) [reply test/extended/cluster/clt.go]: My suggestion was more to make it into a constant to help folks looking at the code understand what 1024 means here.  1024 may well be fine.

File: test/extended/cluster/clt.go [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+
+const DefaultThreadCount int = 10
+
+var _ = g.Describe("[Feature:Performance][Serial][Fast] Load cluster", func() {
+	defer g.GinkgoRecover()
+
+	g.BeforeEach(func() {
+		var err error
+		viperConfig := e2e.TestContext.Viper
+		e2e.Logf("Using config %v\n", viperConfig)
```
Comment (jmencak) [inline test/extended/cluster/clt.go]: `\n"`

Comment (jmencak) [inline test/extended/cluster/clt.go]: `\n"` doesn't seem necessary here.

File: test/extended/cluster/clt.go [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+			e2e.Failf("Unsupported ifexists value '%v' for project %v", p.IfExists, p)
+		}
+
+		if p.IfExists == IF_EXISTS_REUSE && projectExists {
+			// do nothing
+		} else {
+			// Create namespaces as defined in Cluster Loader config
+			err := oc.Run("adm", "new-project").Args(allArgs...).Execute()
+			if err != nil {
+				e2e.Logf("Error creating project: %v\n", err)
```
Comment (jmencak) [inline test/extended/cluster/clt.go]: `\n"` doesn't seem necessary here.  Consistency.

File: test/extended/cluster/clt.go [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+			} else {
+				e2e.Logf("Created new namespace: %v", nsName)
+			}
+		}
+
+		// Create templates as defined
+		for _, template := range p.Templates {
+			err := CreateSimpleTemplates(oc, nsName, e2e.TestContext.Viper, template)
+			if err != nil {
+				e2e.Logf("Error creating template: %v\n", err)
```
Comment (jmencak) [inline test/extended/cluster/clt.go]: `\n"` doesn't seem necessary here. Consistency.

File: test/extended/cluster/clt.go [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+var _ = g.Describe("[Feature:Performance][Serial][Fast] Load cluster", func() {
+	defer g.GinkgoRecover()
+
+	g.BeforeEach(func() {
+		var err error
+		viperConfig := e2e.TestContext.Viper
+		e2e.Logf("Using config %v\n", viperConfig)
+		err = ParseConfig(viperConfig, false)
+		if err != nil {
+			e2e.Failf("Error parsing config: %v\n", err)
```
Comment (jmencak) [inline test/extended/cluster/clt.go]: `\n"` doesn't seem necessary here. Consistency.

Comment (sjug) [reply test/extended/cluster/clt.go]: Fixed

File: test/extended/cluster/clt.go (hunk 6 above)
Comment (sjug) [reply test/extended/cluster/clt.go]: Fixed

File: test/extended/cluster/clt.go (hunk 5 above)
Comment (sjug) [reply test/extended/cluster/clt.go]: Fixed

File: test/extended/cluster/clt.go (hunk 4 above)
Comment (sjug) [reply test/extended/cluster/clt.go]: Fixed

File: test/extended/cluster/clt.go (hunk 2 above)
Comment (sjug) [reply test/extended/cluster/clt.go]: Fixed

**Comment by sjug** [comment]
I think the return exit code 1 is due to my creating clients in goroutines

```
var oc = exutil.NewCLIWithoutNamespace("cl")
```

As inside `NewCLIWithoutNamespace()` there's a `.AfterEach(client.TeardownProject)` which throws the error:

```
fail [github.com/openshift/origin/test/extended/util/cli.go:106]: You may only call AfterEach from within a Describe or Context
```

File: test/extended/cluster/clt.go [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+})
+
+func clWorker(in <-chan ProjectMeta, wg *sync.WaitGroup) {
+	for {
+		p, ok := <-in
+		if !ok {
+			e2e.Logf("Channel closed")
+			return
+		}
+		var oc = exutil.NewCLIWithoutNamespace("cl")
```
Comment (deads2k) [inline test/extended/cluster/clt.go]: pass this as an argument.

Comment (deads2k) [inline test/extended/cluster/clt.go]: if you need it threadsafe, you can make it threadsafe.

Comment (sjug) [reply test/extended/cluster/clt.go]: Updated `SetNamespace()` to fix threading issue.

**Comment by sjug** [comment]
New `SetNamespace()` copy fails govet...
```
# github.com/openshift/origin/test/extended/util
test/extended/util/client.go:187:22: assignment copies lock value to newKubeFramework: github.com/openshift/origin/vendor/k8s.io/kubernetes/test/e2e/framework.Framework contains sync.WaitGroup contains sync.noCopy
[FATAL] FAILURE: go vet failed!
```

**Comment by sjug** [comment]
/unassign @deads2k 

Will try to create templates without the oc command to avoid these client issues, and make a new small PR for just the `SetNamespace()` change.

File: test/extended/cluster/clt.go [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+
+var _ = g.Describe("[Feature:Performance][Serial][Fast] Load cluster", func() {
+	defer g.GinkgoRecover()
+	var oc = exutil.NewCLIWithoutNamespace("cl")
+
+	g.BeforeEach(func() {
+		var err error
+		viperConfig = reale2e.GetViperConfig()
+		e2e.Logf("Using config %v", viperConfig)
+		err = ParseConfig(viperConfig, false)
```
Comment (deads2k) [inline test/extended/cluster/clt.go]: why are you doing this?

File: test/extended/cluster/clt.go [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+var viperConfig string
+
+const (
+	DefaultThreadCount int = 10
+	ChannelSize        int = 1024
+)
+
+var _ = g.Describe("[Feature:Performance][Serial][Fast] Load cluster", func() {
+	defer g.GinkgoRecover()
+	var oc = exutil.NewCLIWithoutNamespace("cl")
```
Comment (deads2k) [inline test/extended/cluster/clt.go]: don't be stingy. What is this client for?

File: test/extended/cluster/clt.go (hunk 9 above)
Comment (sjug) [reply test/extended/cluster/clt.go]: To load user configuration like [this one](https://github.com/openshift/svt/blob/master/openshift_scalability/config/golang/nodeVertical.yaml) (any many others) that we use to drive the test.

File: test/extended/cluster/clt.go (hunk 10 above)
Comment (sjug) [reply test/extended/cluster/clt.go]: That client is the one used to process and create templates

File: test/extended/cluster/context.go [hunk 11]
```diff
@@ -17,12 +17,19 @@ const (
...
 	ClusterLoader struct {
 		Cleanup    bool
+		Threads    int
 		Projects   []ClusterLoaderType
 		Sync       SyncObjectType  `yaml:",omitempty"`
 		TuningSets []TuningSetType `yaml:",omitempty"`
 	}
 }
 
+type ProjectMeta struct {
```
Comment (deads2k) [inline test/extended/cluster/context.go]: If the intent is to expose these as flags through an `openshift-tests` command, this seems weird.

Comment (sjug) [reply test/extended/cluster/context.go]: Initially I was just passing `ClusterLoaderType` so the `clWorker` thread would know what to create, but then I created a custom type to wrap some more fields I needed.

File: test/extended/cluster/utils.go [hunk 12]
```diff
@@ -550,10 +554,93 @@ func CreateTemplates(oc *exutil.CLI, c kclientset.Interface, nsName, config stri
...
+	parameters, ok := temp.Object["parameters"]
+	e2e.Logf("Parameters (%t): %v", ok, parameters)
+	params := parameters.([]interface{})
+	for i, p := range params {
+		e2e.Logf("Parameter index: \"%v\", value: \"%v\"", i, p)
+		field := p.(map[string]interface{})
+		e2e.Logf("Name: %v, Value: %v", field["name"], field["value"])
+		if field["name"] == "IDENTIFIER" {
+			field["value"] = "2"
+			var f interface{} = field
```
Comment (vbatts) [inline test/extended/cluster/utils.go]: since these values are pointers, this replacement is not needed

Comment (sjug) [reply test/extended/cluster/utils.go]: Thanks!

**Pull request merged**